File: src/javascript/algorithms/interval/meetingRooms2.js
//Two pointer TC: O(n logn + n) SC: O(n)

function minMeetingRooms1(intervals) {
    let startTimes = Array(intervals.length);
    let endTimes = Array(intervals.length);

    for(let [start, end] of intervals) {
        startTimes.push(start);
        endTimes.push(end);
    }

    startTimes.sort((a, b) => a-b);
    endTimes.sort((a, b) => a-b);

    let startTimeIndex = endTimeIndex = 0;
    let minMeetingRooms = count = 0; 

    while(startTimeIndex < intervals.length) {
        let startTime = startTimes[startTimeIndex];
        let endTime = endTimes[endTimeIndex];
        if(startTime < endTime) {
            startTimeIndex++;
            count++;
        } else {
            endTimeIndex++;
            count--;
        }
        minMeetingRooms = Math.max(minMeetingRooms, count);
    }
    return minMeetingRooms;
}

let intervals = [[2, 7], [3, 5], [3, 9], [5, 12], [8, 15], [9, 14]];
console.log(minMeetingRooms1(intervals));

File: src/java1/algorithms/interval/meetingRooms2/meetingRooms2.java
package java1.algorithms.interval.meetingRooms2;
import java.util.*;

class Interval {
    int start, end;
    Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
}
public class meetingRooms2 {
    //Two pointer TC: O(n logn) + O(n) ~= O(n), SC: O(n)
    private static int minMeetingRooms1(List<Interval> intervals) {
        int length = intervals.size();
        int[] startTimes = new int[length];
        int[] endTimes = new int[length];
        int minMeetingRooms = 0, count = 0;

        for(int i=0; i<intervals.size(); i++) {
            startTimes[i] = intervals.get(i).start;
            endTimes[i] = intervals.get(i).end;
        }

        Arrays.sort(startTimes);
        Arrays.sort(endTimes);

        int startIndex = 0, endIndex = 0;

        while(startIndex < intervals.size()) {
            if(startTimes[startIndex] < endTimes[endIndex]) {
                count++;
                startIndex++;
            } else {
                count--;
                endIndex++;
            }
            minMeetingRooms = Math.max(minMeetingRooms, count);
        }
        return minMeetingRooms;
    }

    //Priority Queue + sorting:- TC: O(n logn + n) SC: O(n)
    private static int minMeetingRooms2(List<Interval> intervals) {
        Collections.sort(intervals, (a, b) -> a.start - b.start);
        PriorityQueue<Interval> priorityQueue = new PriorityQueue<>((a, b) -> a.end-b.end);
        priorityQueue.add(intervals.get(0));

        for(int i=1; i< intervals.size(); i++) {
            int endTime = priorityQueue.peek().end;
            int startTime = intervals.get(i).start;
            if(endTime <= startTime) {
                priorityQueue.poll();
            }
            priorityQueue.add(intervals.get(i));
        }

        return priorityQueue.size();
    }

    public static void main(String[] args) {
        List<Interval> intervals1 = new ArrayList<>();
        intervals1.add(new Interval(2, 7));
        intervals1.add(new Interval(3, 5));
        intervals1.add(new Interval(3, 9));
        intervals1.add(new Interval(5, 12));
        intervals1.add(new Interval(8, 15));
        intervals1.add(new Interval(9, 14));

        List<Interval> intervals2 = new ArrayList<>();
        intervals2.add(new Interval(1, 5));

        System.out.println(minMeetingRooms1(intervals1));
        System.out.println(minMeetingRooms2(intervals2));
    }
}
